mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -557,6 +557,75 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
   }
 };
 
+// In Tosa canonicalize, a transpose of NCHW to NHWC where H==W==1 will convert
+// to a reshape because it does not change memory layout. Then in TosaToTensor
+// conversion, the reshape is replaced by this pattern:
+//     %0 = collapse(filters[KCHW]) -> [KC]
+//     %1 = expand(%0[KC]) -> [KHWC]
+// If this feeds into a conv2d as filter, we will drop the collapse/expand and
+// update the filter_layout attribute.
+struct CollapseExpandRewritePattern : public OpRewritePattern<tensor::ExpandShapeOp> {
+  using OpRewritePattern<tensor::ExpandShapeOp>::OpRewritePattern;
+
+  bool checkExpand(tensor::ExpandShapeOp expOp) const {
+    auto srcSh = expOp.getOperand().getType().cast<ShapedType>().getShape();
+    auto resSh = expOp.getResultType().cast<ShapedType>().getShape();
+    // [[0, 1, 2], [3]]
+    // NC -> NHWC
+    if (srcSh.size() == 2 && resSh.size() == 4 &&
+        srcSh[0] == resSh[0] && srcSh[1] == resSh[3] &&
+        resSh[1] == 1 && resSh[2] == 1) {
+      return true;
+    }
+    return false;
+  }
+
+  bool checkCollapse(tensor::CollapseShapeOp colOp) const {
+    auto srcSh = colOp.getOperand().getType().cast<ShapedType>().getShape();
+    auto resSh = colOp.getResultType().cast<ShapedType>().getShape();
+    // [[0], [1, 2, 3]]
+    // NCHW -> NC
+    if (srcSh.size() == 4 && resSh.size() == 2 &&
+        srcSh[0] == resSh[0] && srcSh[1] == resSh[1] &&
+        srcSh[2] == 1 && srcSh[3] == 1) {
+      return true;
+    }
+    return false;
+  }
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp expOp,
+                                PatternRewriter &b) const final {
+    LogicalResult lres = failure();
+    
+    Value expInp = expOp.getOperand();
+    Value expOut = expOp.getResult();
+
+    if (!checkExpand(expOp))
+      return failure();
+
+    auto colOp = expInp.getDefiningOp<tensor::CollapseShapeOp>();
+    if (colOp && checkCollapse(colOp)) {
+      auto colInp = colOp.getOperand();
+
+      for (Operation *usr : expOut.getUsers()) {
+        if (auto convOp = dyn_cast<tosa::Conv2DOp>(usr)) {
+          if (convOp.getOperand(1) == expOut) {
```

Comment:
Do we want this pattern for input and output tensors too?

Reply:
Limiting to filters since they are probably the only 1x1 tensors we will see.